### PR TITLE
Cache potential energy and gradient of last trace in HMC/NUTS

### PR DESCRIPTION
**Pull request by neerajprad**
Minor optimization as a result of investigating #1511. 

Since the log density and gradient computation can take much longer with traces containing discrete sites, it is best to avoid it whenever possible. This includes a minor optimization - i.e. caching the potential energy and grads of the last seen trace (we already do caching within the same kernel `sample` step but not across steps).

**Perf Impact:**
The benefits are typically higher for HMC than NUTS, because NUTS tends to build longer chains and therefore the impact of caching the potential energy and gradient values for the beginning of the chain can be small. On the HMM example with 30 time steps with NUTS, this results in a reduction in runtime from 3min 28 seconds to 3min 19 seconds.

### CODE REVIEW

File: pyro/ops/integrator.py [hunk 1]
```diff
@@ -56,30 +46,38 @@ def single_step_velocity_verlet(z, r, potential_fn, inverse_mass_matrix, step_si
...
+        r[site_name] = r[site_name] + 0.5 * step_size * (-z_grads[site_name])  # r(n+1)
+
+    return z, r, z_grads, potential_energy
 
 
 def _potential_grad(potential_fn, z):
     z_keys, z_nodes = zip(*z.items())
     for node in z_nodes:
-        node.requires_grad = True
+        node.requires_grad_(True)
```
Comment (fehiepsi) [inline pyro/ops/integrator.py]: +1

File: pyro/ops/integrator.py [hunk 2]
```diff
@@ -56,30 +46,38 @@ def single_step_velocity_verlet(z, r, potential_fn, inverse_mass_matrix, step_si
     """
     z_next = z.copy()
     r_next = r.copy()
-    z_grads = _potential_grad(potential_fn, z_next)[0] if z_grads is None else z_grads
+    return _single_step_inplace_verlet(z_next, r_next, potential_fn, inverse_mass_matrix, step_size, z_grads)
```
Comment (fehiepsi) [inline pyro/ops/integrator.py]: nit: we might call `velocity_verlet` with `num_steps=1` here

Comment (neerajprad) [reply pyro/ops/integrator.py]: Good point. Yeah, from what I can see, we now don't need `single_step_velocity_verlet`; will refactor.

**Comment by neerajprad** [comment]
This is good to merge, pending any further comments.

**Comment by neerajprad** [comment]
@jpchen - could you help merge this PR (reviewed by @fehiepsi)? Feel free to review too!
